src/test/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporterTest.java

```diff
@@ -1,58 +1,226 @@
-    assertLoggedInCount(0);
-  }
-
-  @Test
-  public void shouldCorrectlyCountImpersonate() throws IOException {
-    final Event imporsonate = createEvent(EventType.IMPERSONATE);
-    PrometheusExporter.instance().recordUserLogin(imporsonate);
-    assertLoggedInCount(1);
-    final Event logout = createEvent(EventType.LOGOUT);
-    PrometheusExporter.instance().recordUserLogout(logout);
-    assertLoggedInCount(0);
-  }
-
-  private void assertLoggedInCount(double number) throws IOException {
-    try (ByteArrayOutputStream stream = new ByteArrayOutputStream()) {
-      PrometheusExporter.instance().export(stream);
-      String result = new String(stream.toByteArray());
-      MatcherAssert.assertThat("Logout count",
-        result.contains("kc_logged_in_users{realm=\"myrealm\",} " + number));
-    }
-  }
-}
\ No newline at end of file
+
+    private static final String DEFAULT_REALM = "myrealm";
+
+    @Before
+    public void before() {
+        for (Counter counter : PrometheusExporter.instance().counters.values()) {
+            counter.clear();
+        }
+        PrometheusExporter.instance().totalLogins.clear();
+        PrometheusExporter.instance().totalFailedLoginAttempts.clear();
+        PrometheusExporter.instance().totalRegistrations.clear();
+    }
+
+    @Test
+    public void shouldRegisterCountersForAllKeycloakEvents() {
+        int userEvents = EventType.values().length;
+        int adminEvents = OperationType.values().length;
+
+        MatcherAssert.assertThat(
+            "All events registered",
+            userEvents + adminEvents - 3,                             // -3 comes from the events that
+            is(PrometheusExporter.instance().counters.size()));       // have their own counters outside the counter map
+
+    }
+
+    @Test
+    public void shouldCorrectlyCountLoginWhenIdentityProviderIsDefined() throws IOException {
+        final Event login1 = createEvent(EventType.LOGIN, tuple("identity_provider", "THE_ID_PROVIDER"));
+        PrometheusExporter.instance().recordLogin(login1);
+        assertMetric("keycloak_logins", 1, tuple("provider", "THE_ID_PROVIDER"));
+
+        final Event login2 = createEvent(EventType.LOGIN, tuple("identity_provider", "THE_ID_PROVIDER"));
+        PrometheusExporter.instance().recordLogin(login2);
+        assertMetric("keycloak_logins", 2, tuple("provider", "THE_ID_PROVIDER"));
+    }
+
+    @Test
+    public void shouldCorrectlyCountLoginWhenIdentityProviderIsNotDefined() throws IOException {
+        final Event login1 = createEvent(EventType.LOGIN);
+        PrometheusExporter.instance().recordLogin(login1);
+        assertMetric("keycloak_logins", 1, tuple("provider", "keycloak"));
+
+        final Event login2 = createEvent(EventType.LOGIN);
+        PrometheusExporter.instance().recordLogin(login2);
+        assertMetric("keycloak_logins", 2, tuple("provider", "keycloak"));
+    }
+
+    @Test
+    public void shouldCorrectlyCountLogin() throws IOException {
```

Comment:
IMO this one should be `shouldCorrectlyCountLoginsFromDifferentProviders` since this case (as it is described) is already cover in the previous 2.